pd-client/client.go

```diff
@@ -14,106 +14,339 @@
+		Header: c.requestHeader(),
+		Count:  uint32(pendingCount + 1),
+	})
+	cancel()
 	requestDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds())
+	if err == nil && resp.GetCount() != uint32(pendingCount+1) {
+		err = errTSOLength
+	}
+	if err != nil {
+		c.finishTSORequest(first, 0, 0, errors.Trace(err))
+		for i := 0; i < pendingCount; i++ {
+			c.finishTSORequest(<-c.tsoRequests, 0, 0, errors.Trace(err))
+		}
+		return
+	}
+
+	physical, logical := resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical()
+	// Server returns the highest ts.
+	logical -= int64(resp.GetCount() - 1)
+	c.finishTSORequest(first, physical, logical, nil)
+	for i := 0; i < pendingCount; i++ {
+		logical++
+		c.finishTSORequest(<-c.tsoRequests, physical, logical, nil)
+	}
+}
+
+func (c *client) finishTSORequest(req *tsoRequest, physical, logical int64, err error) {
+	req.physical, req.logical = physical, logical
+	req.done <- err
+}
+
+func (c *client) Close() {
+	c.cancel()
+	c.wg.Wait()
+
+	n := len(c.tsoRequests)
+	for i := 0; i < n; i++ {
+		req := <-c.tsoRequests
+		req.done <- errors.Trace(errClosing)
+	}
+	for _, cc := range c.connMu.clientConns {
+		if err := cc.Close(); err != nil {
+			log.Errorf("[pd] failed close grpc clientConn: %v", err)
+		}
+	}
+}
+
+func (c *client) leaderClient() pdpb2.PDClient {
+	c.connMu.RLock()
+	defer c.connMu.RUnlock()
+
+	return pdpb2.NewPDClient(c.connMu.clientConns[c.connMu.leader])
+}
 
-	return req.physical, req.logical, err
+func (c *client) scheduleCheckLeader() {
+	select {
+	case c.checkLeaderCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *client) GetClusterID(context.Context) uint64 {
+	return c.clusterID
 }
 
-func (c *client) GetRegion(key []byte) (*metapb.Region, *metapb.Peer, error) {
-	req := &regionRequest{
-		pbReq: &pdpb.GetRegionRequest{
-			RegionKey: key,
-		},
+func (c *client) GetTS(ctx context.Context) (int64, int64, error) {
+	start := time.Now()
+	defer func() { cmdDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds()) }()
+
+	req := &tsoRequest{
 		done: make(chan error, 1),
 	}
+	c.tsoRequests <- req
+
+	select {
+	case err := <-req.done:
+		if err != nil {
+			cmdFailedCounter.WithLabelValues("tso").Inc()
+			c.scheduleCheckLeader()
+			return 0, 0, errors.Trace(err)
+		}
+		return req.physical, req.logical, err
+	case <-ctx.Done():
+		return 0, 0, errors.Trace(ctx.Err())
+	}
+}
 
+func (c *client) GetRegion(ctx context.Context, key []byte) (*metapb.Region, *metapb.Peer, error) {
 	start := time.Now()
-	c.worker.requests <- req
-	err := <-req.done
+	defer func() { cmdDuration.WithLabelValues("get_region").Observe(time.Since(start).Seconds()) }()
+	ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+	resp, err := c.leaderClient().GetRegion(ctx, &pdpb2.GetRegionRequest{
+		Header:    c.requestHeader(),
+		RegionKey: key,
+	})
 	requestDuration.WithLabelValues("get_region").Observe(time.Since(start).Seconds())
```

Comment:
I know the requestDuration and cmdDuration is the same for _get_region_ command. But for _tso_ command, they are different. Thereby, the 2 metrics are both kept.